test/cpp/tensorexpr/test_reductions.cpp

```diff
@@ -0,0 +1,570 @@
+  float expected = 0;
+  for (int i = 0; i < M; ++i) {
+    expected += i;
+  }
+
+  for (int i = 0; i < 2 * 3; ++i) {
+    ASSERT_EQ(cData[i], expected);
+  }
+
+  Tensor* d = Reduce("sum2", {{2, "l"}}, Sum(), b, {{3, "n"}, {m, "m"}});
+  LoopNest loop2({d});
+  loop2.prepareForCodegen();
+  Stmt* s2 = loop2.root_stmt();
+  s2 = IRSimplifier::simplify(s2);
+
+  SimpleIREvaluator cg2(s2, {b, d, m});
+  cg2.call({bData, dData, M});
+
+  // We're combining an additional dimension of 3, so the sum is 3x.
+  expected = expected * 3;
+
+  for (int i = 0; i < 2; ++i) {
+    ASSERT_EQ(dData[i], expected);
+  }
+
+  // This is the same as just reducing the original result across that axis.
+  Buffer c_buf(BufHandle(c->func_var()), kFloat);
+  Tensor* e = Reduce("sum3", {{2, "l"}}, Sum(), c_buf, {{3, "m"}});
+  LoopNest loop3({e});
+  loop3.prepareForCodegen();
+  Stmt* s3 = loop3.root_stmt();
+  s3 = IRSimplifier::simplify(s3);
+
+  SimpleIREvaluator cg3(s3, {c, e});
+  cg3.call({cData, eData});
+
+  for (int i = 0; i < 2; ++i) {
+    ASSERT_EQ(eData[i], expected);
+  }
+}
+
+// Sum a large (10 D) Tensor 5 dimensions in.
+void testReduceSum10D() {
+  KernelScope kernel_scope;
+
+  Buffer in_(BufHandle("in_", {2, 3, 2, 3, 2, 3, 2, 3, 2, 3}), kFloat);
+  const int InputSize = 2 * 3 * 2 * 3 * 2 * 3 * 2 * 3 * 2 * 3;
+  Buffer out_(BufHandle("out_", {2, 3, 2, 3, 2}), kFloat);
+  const int OutputSize = 2 * 3 * 2 * 3 * 2;
+
+  std::vector<float> in(InputSize, 1.f);
+  std::vector<float> out(OutputSize, -1.f);
+
+  Tensor* c = Reduce(
+      "sum",
+      {{2, "a"}, {3, "b"}, {2, "c"}, {3, "d"}, {2, "e"}},
+      Sum(),
+      in_,
+      {{3, "f"}, {2, "g"}, {3, "h"}, {2, "i"}, {3, "j"}});
+  LoopNest loop({c});
+  loop.prepareForCodegen();
+  Stmt* s = loop.root_stmt();
+  s = IRSimplifier::simplify(s);
+
+  SimpleIREvaluator cg(s, {in_, c});
+
+  cg.call({in, out});
+
+  float expected = InputSize / OutputSize;
+  for (int i = 0; i < OutputSize; ++i) {
+    ASSERT_EQ(out[i], expected);
+  }
+}
+
+// Reduce via Mul rather than Add using a custom Reducer.
+void testReduceProduct() {
+  KernelScope kernel_scope;
+
+  const int M = 4;
+  const int N = 4;
+
+  Buffer b(BufHandle("b", {M, N}), kFloat);
+  std::vector<float> in(M * N);
+  for (int i = 0; i < M; ++i) {
+    for (int j = 0; j < N; ++j) {
+      in[i * N + j] = 2 + j;
+    }
+  }
+
+  std::vector<float> out(M, -1.f);
+
+  Reducer product(
+      ExprHandle(1.f), [](ExprHandle a, ExprHandle b) { return a * b; });
+
+  Tensor* c = Reduce("product", {{M, "m"}}, product, b, {{N, "n"}});
+  LoopNest loop({c});
+  loop.prepareForCodegen();
+  Stmt* s = loop.root_stmt();
+  s = IRSimplifier::simplify(s);
+
+  SimpleIREvaluator cg(s, {b, c});
+
+  cg.call({in, out});
+
+  float expected = 1;
+  for (int i = 0; i < N; ++i) {
+    expected *= 2 + i;
+  }
+
+  for (int i = 0; i < M; ++i) {
+    ASSERT_EQ(out[i], expected);
+  }
+}
+
+// Foldl of a noncommutative operator (Sub) from a inital value.
```

Comment:
[TODO] would prefer if fold was separate (as it induces an ordering that's hard to optimize)

Reply:
What do you mean by separate? Not as a part of the reduce call?

Reply:
I now understand, I had assumed For implied an ordering but since it doesn't this case isn't expressible. I'll remove the test but if somebody wants to write a custom reducer which assumes ordering im not sure how we stop them.